nvtabular/ops/normalize.py

```diff
@@ -46,24 +47,31 @@ def __init__(self):
         super().__init__()
         self.means = {}
         self.stds = {}
+        self.means_cupy = self.stds_cupy = None
 
     @annotate("Normalize_fit", color="green", domain="nvt_python")
     def fit(self, columns: ColumnNames, ddf: dask_cudf.DataFrame):
         return _custom_moments(ddf[columns])
 
     def fit_finalize(self, dask_stats):
+        self.means_cupy = cupy.array(dask_stats["mean"].values)
+        self.stds_cupy = cupy.array(dask_stats["std"].values)
+
         for col in dask_stats.index:
             self.means[col] = float(dask_stats["mean"].loc[col])
             self.stds[col] = float(dask_stats["std"].loc[col])
 
     @annotate("Normalize_op", color="darkgreen", domain="nvt_python")
     def transform(self, columns: ColumnNames, gdf: cudf.DataFrame) -> cudf.DataFrame:
-        new_gdf = cudf.DataFrame()
-        for name in columns:
-            if self.stds[name] > 0:
-                new_gdf[name] = (gdf[name] - self.means[name]) / (self.stds[name])
-                new_gdf[name] = new_gdf[name].astype("float32")
-        return new_gdf
+        arr = cupy.fromDlpack(gdf[columns].astype("float32").to_dlpack())
```

Comment:
Is there a good reason to do `to_dlpack`/`fromDlpack` instad of using `values` directly? That is, doesn't this produce the same result: `arr = gdf[columns].astype("float32").values`

My preference for the `values` approach (if possible/performant) is that it should work for both `pandas` and `cudf`

Reply:
Thanks for the review!

> Is there a good reason to do to_dlpack/fromDlpack instad of using values directly? That is, doesn't this produce the same result: arr = gdf[columns].astype("float32").values


Yeah-  this code here is just a POC and I don't think we should merge as is (especially with your most recent PR that adds cpu support =).  The motivating idea here is to speed up small batch sizes (for instance when doing inference). This PR leads to about a 4x speedup in the normalize transform time by doing all the columns at once.

I agree that we should make this work with both the cpu/gpu versions - and the 'values' method you proposed is an elegant way of handling that. However, the dlpack version is about 2x faster than the .values version if you include the 'astype' cast - and about 6x faster without it:

```python
df = cudf.DataFrame({"x": [1.,2.,3.,4.,5.], "y":[10.,20.,30.,40., 50.]})

# With converting to float32:

min(timeit.Timer(lambda: df.astype("float32").values).repeat(repeat=100, number=1))
# 0.00040847109630703926
min(timeit.Timer(lambda: cupy.fromDlpack(df.astype("float32").to_dlpack())).repeat(repeat=100, number=1)) 
# 0.000211270060390234

# Without converting to float32:
min(timeit.Timer(lambda: df.values).repeat(repeat=100, number=1))
#  0.00016679614782333374

min(timeit.Timer(lambda: cupy.fromDlpack(df.to_dlpack())).repeat(repeat=100, number=1))
# 4.502898082137108e-05
```
Its *probably* still worth it do as you suggested - the total times here are small - but its something we should keep in mind.

Reply:
Wow - Thats very interesting!  Thanks for clarifying that there is indeed an advantage of dlpack to consider here.